atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -362,31 +371,103 @@ public void putWithTimestamps(TableReference tableRef, Multimap<Cell, Value> val
+            throws KeyAlreadyExistsException {
+        putInternal(tableRef, KeyValueServices.toConstantTimestampValues(values.entrySet(), 0), true);
+    }
+
     private void putInternal(TableReference tableRef, Collection<Map.Entry<Cell, Value>> values, boolean doNotOverwriteWithSameValue) {
         Table table = getTableMap(tableRef);
         for (Map.Entry<Cell, Value> e : values) {
-            Cell cell = e.getKey();
-            byte[] row = cell.getRowName();
-            byte[] col = cell.getColumnName();
             byte[] contents = e.getValue().getContents();
             long timestamp = e.getValue().getTimestamp();
 
-            Key nextKey = table.entries.ceilingKey(new Key(row, ArrayUtils.EMPTY_BYTE_ARRAY, Long.MIN_VALUE));
-            if (nextKey != null && nextKey.matchesRow(row)) {
-                // Save memory by sharing rows.
-                row = nextKey.row;
-            }
-            byte[] oldContents = table.entries.putIfAbsent(new Key(row, col, timestamp), copyOf(contents));
+            Key key = getKey(table, e.getKey(), timestamp);
+            byte[] oldContents = putIfAbsent(table, key, contents);
             if (oldContents != null && (doNotOverwriteWithSameValue || !Arrays.equals(oldContents, contents))) {
                 throw new KeyAlreadyExistsException("We already have a value for this timestamp");
             }
         }
     }
 
+    // Returns the existing contents, if any, and null otherwise
+    private byte[] putIfAbsent(Table table, Key key, final byte[] contents) {
+        return table.entries.putIfAbsent(key, copyOf(contents));
+    }
+
     @Override
-    public void putUnlessExists(TableReference tableRef, Map<Cell, byte[]> values)
-            throws KeyAlreadyExistsException {
-        putInternal(tableRef, KeyValueServices.toConstantTimestampValues(values.entrySet(), 0), true);
+    public void checkAndSet(CheckAndSetRequest request) throws CheckAndSetException {
+        Table table = getTableMap(request.table());
+        Cell cell = request.cell();
+        Optional<byte[]> oldValue = request.oldValue();
+        byte[] contents = request.newValue();
+
+        Key key = getKey(table, cell, AtlasDbConstants.TRANSACTION_TS);
+        if (oldValue.isPresent()) {
+            byte[] storedValue = table.entries.get(key);
+            boolean succeeded = Arrays.equals(storedValue, oldValue.get())
+                    && table.entries.replace(key, storedValue, copyOf(contents));
+            if (!succeeded) {
+                byte[] actual = table.entries.get(key); // Refetch, something may have happened between get and replace
+                throwCheckAndSetException(cell, oldValue.get(), actual);
+            }
+        } else {
+            byte[] oldContents = putIfAbsent(table, key, contents);
+            if (oldContents != null) {
+                String actual = new String(oldContents, StandardCharsets.UTF_8);
```

Comment:
Any reason why we assume the buffer is a UTF_8 string here?  In Cell.java we just dump it into hex since we don't know what these bytes represent.